File: ios/Classes/FfmpegCbsPlugin.swift
import Flutter
import ffmpegkit

public class FFMpegCBSPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private var methodChannel: FlutterMethodChannel?
    private var eventSink: FlutterEventSink?
    private var session: FFmpegSession?

    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "ffmpeg_cbs/method", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "ffmpeg_cbs/event", binaryMessenger: registrar.messenger())
        
        let instance = FFMpegCBSPlugin()
        instance.methodChannel = methodChannel
        
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)

        print("[FFmpegCBS] Plugin registered successfully.")
    }

    // MARK: - Method Call Handling

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[FFmpegCBS] Received method call: \(call.method)")

        switch call.method {
        case "getPlatformVersion":
            let version = "iOS " + UIDevice.current.systemVersion
            print("[FFmpegCBS] Returning platform version: \(version)")
            result(version)

        case "startRecord":
            guard let args = call.arguments as? [String: Any],
                  let command = args["command"] as? String else {
                print("[FFmpegCBS] startRecord: Invalid arguments received.")
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing command", details: nil))
                return
            }
            print("[FFmpegCBS] Starting recording with command:\n\(command)")
            startRecording(with: command)
            result(nil)

        case "stopRecord":
            print("[FFmpegCBS] Stopping recording session.")
            stopRecording()
            result(nil)

        default:
            print("[FFmpegCBS] Method not implemented: \(call.method)")
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FFmpeg Operations

    private func startRecording(with command: String) {
        print("[FFmpegCBS] FFmpeg session started")

        session = FFmpegKit.executeAsync(command, withCompleteCallback: self.handleFFmpegCompletion,
            withLogCallback: { log in
            if let log = log {
                print("[FFmpegCBS][Log] \(log.getSessionId()): lvl\(log.getLevel()) \(log.getMessage().description)")
            }
        }, withStatisticsCallback: { statistics in
            
        })
    }

    private func stopRecording() {
        print("[FFmpegCBS] Cancelling FFmpegKit session.")
        guard let session = self.session else {
            print("[FFmpegCBS] No active session to cancel.")
            return
        }
        session.cancel()
        FFmpegKit.cancel()
    }
    
    private func handleFFmpegCompletion(session: FFmpegSession?) {
        guard let session = session else {
            print("[FFmpegCBS] Session is nil")
            return
        }

        let returnCode = session.getReturnCode()
        let output = session.getOutput() ?? "(no output)"
        let failStackTrace = session.getFailStackTrace()

        // 시간 정보
        let startTime = session.getStartTime()
        let endTime = session.getEndTime()
        let durationSec = Double(session.getDuration()) / 1000.0

        print("[FFmpegCBS] Start time: \(startTime?.description ?? "nil") End time: \(endTime?.description ?? "nil")")
        print("[FFmpegCBS] Duration: \(durationSec) sec")

        print("[FFmpegCBS] FFmpeg session completed. ReturnCode: \(returnCode?.description ?? "Unknown")")
        print("[FFmpegCBS] Session output:\n\(output)")
        
        var message = [:] as [String: Any]

        if ReturnCode.isSuccess(returnCode) {
            print("[FFmpegCBS] Recording SUCCESS.")
            message = ["status": "completed",
                       "duration": durationSec]
        } else if ReturnCode.isCancel(returnCode) {
            print("[FFmpegCBS] Recording CANCELLED.")
            message = ["status": "cancelled",
                       "duration": durationSec]
        } else {
            print("[FFmpegCBS] Recording FAILED.")
            message["status"] = "error"
            message["error"] = failStackTrace ?? "Unknown error"
        }
        DispatchQueue.main.async {
            guard let sink = self.eventSink else {
                print("[FFmpegCBS] EventSink is nil, cannot send recording status.")
                return
            }
            sink(message)
       }
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("[FFmpegCBS] Event channel: onListen")
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        print("[FFmpegCBS] Event channel: onCancel")
        self.eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/kr/co/cbs/ffmpeg/FFMpegCBSPlugin.kt
package com.example.ffmpeg_cbs

import android.os.Build
import android.util.Log
import com.arthenica.ffmpegkit.FFmpegKit
import com.arthenica.ffmpegkit.FFmpegSession
import com.arthenica.ffmpegkit.ReturnCode
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

/** FFMpegCBSPlugin */
class FFMpegCBSPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private var eventSink: EventChannel.EventSink? = null
  private var session: FFmpegSession? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "ffmpeg_cbs/method")
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "ffmpeg_cbs/event")

    methodChannel.setMethodCallHandler(this)
    eventChannel.setStreamHandler(this)
    Log.d("FFMpegCBS", "Plugin registered successfully")
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${Build.VERSION.RELEASE}")
      }
      "startRecord" -> {
        val command = call.argument<String>("command")
        if (command == null) {
          Log.e("FFMpegCBS", "Invalid arguments for startRecord")
          result.error("INVALID_ARGUMENT", "Missing command", null)
          return
        }
        Log.d("FFMpegCBS", "Starting recording with command: $command")
        startRecording(command)
        result.success(null)
      }
      "stopRecord" -> {
        Log.d("FFMpegCBS", "Stopping recording session")
        stopRecording()
        result.success(null)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun startRecording(command: String) {
    session = FFmpegKit.executeAsync(command,
      { session -> handleFFmpegCompletion(session) },
      { log ->
        log?.let {
          Log.d("FFMpegCBS", "[Log] ${it.sessionId}: lvl${it.level} ${it.message}")
        }
      },
      { statistics ->
        // optional stats handling
      }
    )
  }

  private fun stopRecording() {
    session?.cancel()
    FFmpegKit.cancel()
  }

  private fun handleFFmpegCompletion(session: FFmpegSession?) {
    if (session == null) {
      Log.e("FFMpegCBS", "Session is null")
      return
    }

    val returnCode = session.returnCode
    val output = session.output ?: "(no output)"
    val failStackTrace = session.failStackTrace

    val startTime = session.startTime
    val endTime = session.endTime
    val durationSec = session.duration.toDouble() / 1000.0

    Log.d("FFMpegCBS", "Start: $startTime, End: $endTime, Duration: $durationSec sec")
    Log.d("FFMpegCBS", "ReturnCode: ${returnCode?.value} Output: $output")

    val message = mutableMapOf<String, Any>()
    when {
      ReturnCode.isSuccess(returnCode) -> {
        Log.d("FFMpegCBS", "Recording SUCCESS")
        message["status"] = "completed"
        message["duration"] = durationSec
      }
      ReturnCode.isCancel(returnCode) -> {
        Log.d("FFMpegCBS", "Recording CANCELLED")
        message["status"] = "cancelled"
        message["duration"] = durationSec
      }
      else -> {
        Log.d("FFMpegCBS", "Recording FAILED")
        message["status"] = "error"
        message["error"] = failStackTrace ?: "Unknown error"
      }
    }

    eventSink?.success(message)
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    Log.d("FFMpegCBS", "Event channel: onListen")
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    Log.d("FFMpegCBS", "Event channel: onCancel")
    eventSink = null
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }
}